api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,268 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of context nodes which each have access to the nodes before it, but not
+// after. So each time a write is performed a new context is added to the list.
+// Within each context node there is another linked list which holds the key
+// and value data for each node.
+class Context
+{
+
+public:
+  nostd::shared_ptr<Context> *context_ptr_;
+
+  // Creates a new context by calling a constructor with the passed in
+  // keyValueIterable and returns a shared_ptr pointing to it.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  static nostd::shared_ptr<Context> CreateContext(T &keys_and_values)
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(keys_and_values)};
+    temp->context_ptr_              = &temp;
+    return temp;
+  }
+
+  // Creates a new context by calling a constructor with the passed in
+  // key and value and returns a shared_ptr pointing to it.
+  static nostd::shared_ptr<Context> CreateContext(nostd::string_view key, ContextValue value)
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(key, value)};
+    temp->context_ptr_              = &temp;
+    return temp;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  nostd::shared_ptr<Context> SetValues(T &values) noexcept
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(values)};
+    temp->context_ptr_              = &temp;
```

Comment:
I think storing `context_ptr_` like this is dangerous: it points to a `nostd::shared_ptr` on the stack, this might get copied around and the address it points to can be invalid at a later point in time.

You could get around the need of storing `context_ptr_` by making `SetValues` methods static, or by taking them out of the class as functions. Then you can pass in a parent as a `nostd::shared_ptr` explicitly:
```c
template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
nostd::shared_ptr<Context> SetValues(nostd::shared_ptr<Context> parent, T &values) noexcept
{
  nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(values)};
  temp->context_ptr_              = &temp;
  temp->SetNext(parent);
  return temp;
}
```

Reply:
I modified the implementation considerably it now avoids the issue, could I convince you to take a look at the code again?